### DISCUSSION TITLE
Whats the trick to adding Google Optimize?

### DISCUSSION THREAD

**Discussion by joshbedo** [vercel/next.js#18991] [Help]
I tried adding the script tag to next/head and have a test running but the variant doesn't change the text on the screen? Can't seem to get a/b testing to work with Google Optimize. Is there something I'm missing? TIA!

**Comment by Hewlbern** [comment]
Working out how to do A/B Testing with next.js myself!

**Reply by brumeregan** [reply]
Hi! Did you find the right solution? I am facing the same problem

**Reply by tomk-** [reply]
Yes, please share @joshbedo if you had any resolution on this!

**Comment by jdorn** [comment]
NextJS serves static or server-rendered html initially and then when the javascript loads, it "hydrates" the page and re-renders with React.  Google Optimize is also trying to modify the HTML right after the initial page load.  This causes a race condition where Google Optimize may change the HTML first but then React hydrates the page and changes it back.

You can load Google Optimize after hydration, but that's not ideal since the user will see a "flash" and it can still break if something else causes your app to re-render.  You can use the [Google Optimize javascript API](https://support.google.com/optimize/answer/9059383) to run experiments programatically, but that doesn't work well for SSR and is pretty tedious to use.

I might recommend just skipping Google Optimize altogether and implementing the A/B tests yourself.  I'm working on a React A/B testing library (https://github.com/growthbook/growthbook-react) to make this a little easier:

```tsx
export default function MyComponent() {
  const {value} = useExperiment({
    key: "my-test",
    variations: ["Control Headline", "Variant Headline 1"]
  });
  return <h1>{value}</h1>
}
```

You can still use Google Analytics to track metrics and report on the data.  It won't do the stats calculations, but that's not too hard to do yourself with online A/B test calculators.

**Comment by ivanfgm** [comment]
I recently implemented an A/B test on a next.js site, with the help of [this article](https://medium.com/broadlume-product/how-to-add-google-optimize-a-b-testing-to-your-react-app-in-10-lines-of-code-8310b58e51f4). The trick is to initialize optimize yourself to avoid race conditions with `window.dataLayer.push({ event: "optimize.activate" })`, after setting the 'Activation event' to 'custom event' in the experiment settings.

As stated, when using the editor to make the variants you'll end up with a flash between changes. However you can just use the experiment variant info in react an decide what to do there, in my case the A/B test controlled whether a modal box appeared on load, or after clicking a button, and for this kind of use case I found it ideal.

Here's what I ended up with, the variant returned by google is a positive int starting in 0 for the original, I didn't particularly like the interval to wait for initialization, but it works fine:

```javascript
export default function useOptimizeAB(experimentId) {
  const [variant, setVariant] = useState();
  useEffect(() => {
    window.dataLayer.push({ event: "optimize.activate" });
    let interval = setInterval(() => {
      if (window.google_optimize !== undefined) {
        const variant = window.google_optimize.get(experimentId);
        if (typeof variant !== "undefined") setVariant(Number(variant));
        clearInterval(interval);
      }
    }, 100);
  }, []);
  return variant;
}

// ... then in _app.js you can use the variant for whatever
const variant = useOptimizeAB("your_experiment_id");
```

**Reply by vaibhav16** [reply]
Hey, I did the exact same steps as mentioned in the article and but for some reason, my variant window.google_optimize.get(experimentId)) is always undefined. I checked my installation in google optimize dashboard and it also says that my installation is correct.

Did you experience any such issue?

**Reply by ayushmamaearth** [reply]
I also experienced the same issue as menttioned above by vaibhav16

**Reply by ivanfgm** [reply]
Hello @vaibhav16, it sure didn't work on the first try, I would double check:
- You have the 'Activation event' option as 'custom event' in the experiment settings.
- The experiment is running and the id is correct.
- You are calling `window.dataLayer.push({ event: "optimize.activate" })` before the interval.

Those are the things I was doing wrong when getting undefined, hope it helps.

**Reply by vaibhav16** [reply]
@marioivangf 
I didn't start the experiment and that is why the variant was undefined.
Thank you for helping out!

@ayushmamaearth This is what solved the issue for me.

**Reply by laamia-alula** [reply]
Does anyone know how to test this implementation locally? Google Optimize makes you specify a target url for your experiment.

**Reply by vaibhav16** [reply]
@laamia-alula 
You just need to change your host settings. 

Copy the address assigned to localhost (typically it should be` 127.0.0.1`) in your hosts file and add it as a new entry with a custom URL. For instance, something like **test.localhost.com**

If your project was earlier working on **localhost:3000**, now it should also work on **test.localhost.com:3000**
Add this new url as the target url in Google Optimize Dashboard. 

That's all there is to it.

**Reply by ayushmamaearth** [reply]
> @marioivangf I didn't start the experiment and that is why the variant was undefined. Thank you for helping out!
> 
> @ayushmamaearth This is what solved the issue for me.

yes @vaibhav16 my issue was also resolved. I made a different setup with the help of google tag manager and redux properties at code level. Thankyou for the help.

**Reply by davidornelas11** [reply]
Could you use the variant in your components as well to dynamically render your tests?

**Reply by adrianburgosh** [reply]
Hi, I've implemented all the steps but when I use the hook and load the page it works but when I navigate to it via my nav bar optimizer didn't set the cookie for the experiment but I don't get it, if anyone has been through this, this is my hook

```javascript
const useExperiment = (experimentId, defaultVariant) => {
	const [variant, setVariant] = useState();
	useEffect(() => {
		if (window.dataLayer) {
			window.dataLayer.push({ event: "optimize.activate" });
		}
		const intervalId = setInterval(() => {
			if (window.google_optimize !== undefined) {
				const variantLoaded = window.google_optimize.get(experimentId);
				setVariant(variantLoaded || defaultVariant);
				clearInterval(intervalId);
			}
		}, 100);
	}, []);
	return variant;
};
```

implementation
```javascript
const variant = useExperiment("<EXPERIMENT_ID>", DEFAULT_VARIANT);
```

SOLUTION
I was missing on the Google optimize configuration to add the custom Activation event with value of "optimize.active" in this case

**Reply by strtw** [reply]
@jdorn , @marioivangf 

I've managed to implement a personalization experiment on a NextJS site which runs the optimize changes based on a custom activation event, but I'm seeing flicker. I've tried adding the [optimize anti-flicker snippet](https://support.google.com/optimize/answer/7100284) above the optimize script to no avail. 

Google says: 
"Do not install the anti-flicker snippet:

...if your Optimize snippet isn't at the top of the <HEAD>.  
...if you haven't observed page flicker.
...if you use [activation events](https://support.google.com/optimize/answer/7008840) that don't fire changes on the first load.
...with a tag management system (TMS) like Google Tag Manager (GTM)."

However, I can't load Optimize at the top of the <HEAD> because 3rd party consent scripts need to load before Optimize via Google Tag Manager. And I am using an activation event that doesn't fire on first load, but on the result of a callback function in the Optimize script. 

Any ideas on how to mitigate the flicker issue?

**Reply by d-spindola** [reply]
Maybe we should start to think more in how to use Dynamic import for this, also, using Script tag from Next we can load the GTM container smoothly. But the next step is enable the Google Optimize, I think that we could achieve this with Dynamic, I'm doing some work now and will share my code with you folks

**Reply by d-spindola** [reply]
Actually I'm responsible to implement A/B test and we use Amplify which has no built-in feature to achieve this, I know that vercel has edge which enables us to perform operations over requests during page rendering, with edge architecture we can rewrite the response and inject cookies to handle content display.
